README.md
# Redirector plugin for Craft CMS 3.x

A simple way to add 301/302 redirects within CraftCMS. This is the first CraftCMS plugin written by myself so please do let me know any bugs you may find.

![Screenshot](https://i.ibb.co/hsd6Pqj/image.png)

## Requirements

This plugin requires Craft CMS 3.0.0-beta.23 or later.

## Installation

To install the plugin, follow these instructions.

1. Open your terminal and go to your Craft project:

        cd /path/to/project

2. Then tell Composer to load the plugin:

        composer require jaetooledev/redirector

3. In the Control Panel, go to Settings → Plugins and click the “Install” button for Redirector.

## Redirector Overview

Are you wanting to implement some 301 and/or 302 redirects within your CraftCMS projects? If so, Redirector provides a simple utility to make that happen. You can redirect to internal and external links and (coming soon) interact with Redirector via GraphQL.

## Redirector Roadmap

Some things to do, and ideas for potential features:

* GraphQL support
* CSV uploading
* Using datatables to display redirects
* Analytics support
* Event emitting
* More translations
* Cleaning codebase
* Adding tests

Brought to you by [Jae Toole](https://github.com/jaetooledev)


CHANGELOG.md
# Redirector Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## 1.0.0 - 2021-11-25
### Added
- Initial release


tests/_support/UnitTester.php
<?php
/**
 * Redirector plugin for Craft CMS 3.x
 *
 * A simple way to add 301/302 redirects within CraftCMS
 *
 * @link      https://github.com/jaetooledev
 * @copyright Copyright (c) 2021 Jae Toole
 */

use Codeception\Actor;
use Codeception\Lib\Friend;

/**
 * Inherited Methods
 *
 * @method void wantToTest($text)
 * @method void wantTo($text)
 * @method void execute($callable)
 * @method void expectTo($prediction)
 * @method void expect($prediction)
 * @method void amGoingTo($argumentation)
 * @method void am($role)
 * @method void lookForwardTo($achieveValue)
 * @method void comment($description)
 * @method Friend haveFriend($name, $actorClass = null)
 *
 * @SuppressWarnings(PHPMD)
 *
 */
class UnitTester extends Actor
{
    use _generated\UnitTesterActions;

}


tests/_support/Helper/Unit.php
<?php
/**
 * Redirector plugin for Craft CMS 3.x
 *
 * A simple way to add 301/302 redirects within CraftCMS
 *
 * @link      https://github.com/jaetooledev
 * @copyright Copyright (c) 2021 Jae Toole
 */

namespace Helper;

use Codeception\Module;

/**
 * Class Unit
 *
 * Here you can define custom actions.
 * All public methods declared in helper class will be available in $I
 *
 */
class Unit extends Module
{

}


tests/unit/ExampleUnitTest.php
<?php
/**
 * Redirector plugin for Craft CMS 3.x
 *
 * A simple way to add 301/302 redirects within CraftCMS
 *
 * @link      https://github.com/jaetooledev
 * @copyright Copyright (c) 2021 Jae Toole
 */

namespace jaetooledev\redirectortests\unit;

use Codeception\Test\Unit;
use UnitTester;
use Craft;
use jaetooledev\redirector\Redirector;

/**
 * ExampleUnitTest
 *
 *
 * @author    Jae Toole
 * @package   Redirector
 * @since     1.0.0
 */
class ExampleUnitTest extends Unit
{
    // Properties
    // =========================================================================

    /**
     * @var UnitTester
     */
    protected $tester;

    // Public methods
    // =========================================================================

    // Tests
    // =========================================================================

    /**
     *
     */
    public function testPluginInstance()
    {
        $this->assertInstanceOf(
            Redirector::class,
            Redirector::$plugin
        );
    }

    /**
     *
     */
    public function testCraftEdition()
    {
        Craft::$app->setEdition(Craft::Pro);

        $this->assertSame(
            Craft::Pro,
            Craft::$app->getEdition()
        );
    }
}


src/Redirector.php
<?php

namespace jaetooledev\redirector;

use jaetooledev\redirector\services\RedirectService;
use Craft;
use craft\base\Plugin;
use craft\services\Plugins;
use craft\events\PluginEvent;
use craft\web\UrlManager;
use craft\events\RegisterUrlRulesEvent;
use yii\base\Event;

class Redirector extends Plugin
{

    /**
     * Static property that is an instance of this plugin class so that it can be accessed via
     * Redirector::$plugin
     *
     * @var Redirector
     */
    public static $plugin;

    public $schemaVersion = '1.0.0';

    public $hasCpSettings = false;

    public $hasCpSection = true;

    public function init()
    {
        parent::init();
        $this->setComponents(
            [
                'redirectService' => RedirectService::class
            ]
        );
        self::$plugin = $this;

        Event::on(
            UrlManager::class,
            UrlManager::EVENT_REGISTER_SITE_URL_RULES,
            function (RegisterUrlRulesEvent $event) {
                $redirects = self::getInstance()->redirectService->getRedirectsForSite(
                    Craft::$app->getSites()->currentSite->id
                );
                $redirectRules = [];
                foreach ($redirects as $redirect) {
                    $redirectRules[$redirect['fromUrl']] = [
                        'route' => 'redirector/redirect/index',
                        'params' => [
                            'redirectId' => $redirect['id']
                        ]
                    ];
                }

                $event->rules += $redirectRules;
            }
        );

        Event::on(
            UrlManager::class,
            UrlManager::EVENT_REGISTER_CP_URL_RULES,
            function (RegisterUrlRulesEvent $event) {
                $event->rules['redirector'] = 'redirector/dashboard/index';
                $event->rules['redirector/redirect/new'] = 'redirector/dashboard/edit';
                $event->rules['redirector/redirect/<redirectId:\d+>'] = 'redirector/dashboard/edit';
                $event->rules['redirector/redirect/<redirectId:\d+>/delete'] = 'redirector/dashboard/delete';
            }
        );

        Craft::info(
            Craft::t(
                'redirector',
                '{name} plugin loaded',
                ['name' => $this->name]
            ),
            __METHOD__
        );
    }

    public function getCpNavItem(): ?array
    {
        $item = parent::getCpNavItem();
        $item['url'] = 'redirector';
        $item['icon'] = '@jaetooledev/redirector/redirector.svg';
        return $item;
    }
}


src/migrations/Install.php
<?php
/**
 * Redirector plugin for Craft CMS 3.x
 *
 * A simple way to add 301/302 redirects within CraftCMS
 *
 * @link      https://github.com/jaetooledev
 * @copyright Copyright (c) 2021 Jae Toole
 */

namespace jaetooledev\redirector\migrations;

use jaetooledev\redirector\Redirector;

use Craft;
use craft\config\DbConfig;
use craft\db\Migration;

/**
 * Redirector Install Migration
 *
 * If your plugin needs to create any custom database tables when it gets installed,
 * create a migrations/ folder within your plugin folder, and save an Install.php file
 * within it using the following template:
 *
 * If you need to perform any additional actions on install/uninstall, override the
 * safeUp() and safeDown() methods.
 *
 * @author    Jae Toole
 * @package   Redirector
 * @since     1.0.0
 */
class Install extends Migration
{
    // Public Properties
    // =========================================================================

    /**
     * @var string The database driver to use
     */
    public $driver;

    // Public Methods
    // =========================================================================

    /**
     * This method contains the logic to be executed when applying this migration.
     * This method differs from [[up()]] in that the DB logic implemented here will
     * be enclosed within a DB transaction.
     * Child classes may implement this method instead of [[up()]] if the DB logic
     * needs to be within a transaction.
     *
     * @return boolean return a false value to indicate the migration fails
     * and should not proceed further. All other return values mean the migration succeeds.
     */
    public function safeUp()
    {
        $this->driver = Craft::$app->getConfig()->getDb()->driver;
        if ($this->createTables()) {
            $this->createIndexes();
            $this->addForeignKeys();
            // Refresh the db schema caches
            Craft::$app->db->schema->refresh();
            $this->insertDefaultData();
        }

        return true;
    }

    /**
     * This method contains the logic to be executed when removing this migration.
     * This method differs from [[down()]] in that the DB logic implemented here will
     * be enclosed within a DB transaction.
     * Child classes may implement this method instead of [[down()]] if the DB logic
     * needs to be within a transaction.
     *
     * @return boolean return a false value to indicate the migration fails
     * and should not proceed further. All other return values mean the migration succeeds.
     */
    public function safeDown()
    {
        $this->driver = Craft::$app->getConfig()->getDb()->driver;
        $this->removeTables();

        return true;
    }

    // Protected Methods
    // =========================================================================

    /**
     * Creates the tables needed for the Records used by the plugin
     *
     * @return bool
     */
    protected function createTables()
    {
        $tablesCreated = false;

    // redirector_redirects table
        $tableSchema = Craft::$app->db->schema->getTableSchema('{{%redirector_redirects}}');
        if ($tableSchema === null) {
            $tablesCreated = true;
            $this->createTable(
                '{{%redirector_redirects}}',
                [
                    'id' => $this->primaryKey(),
                    'dateCreated' => $this->dateTime()->notNull(),
                    'dateUpdated' => $this->dateTime()->notNull(),
                    'uid' => $this->uid(),
                    'siteId' => $this->integer()->notNull(),
                    'fromUrl' => $this->string(255)->notNull()->defaultValue(''),
                    'toUrl' => $this->string(255)->notNull()->defaultValue(''),
                    'type' => $this->integer(3)->notNull()->defaultValue(301)
                ]
            );
        }

        return $tablesCreated;
    }

    /**
     * Creates the foreign keys needed for the Records used by the plugin
     *
     * @return void
     */
    protected function addForeignKeys()
    {
    // redirector_redirects table
        $this->addForeignKey(
            $this->db->getForeignKeyName('{{%redirector_redirects}}', 'siteId'),
            '{{%redirector_redirects}}',
            'siteId',
            '{{%sites}}',
            'id',
            'CASCADE',
            'CASCADE'
        );
    }

    /**
     * Populates the DB with the default data.
     *
     * @return void
     */
    protected function insertDefaultData()
    {
    }

    /**
     * Removes the tables needed for the Records used by the plugin
     *
     * @return void
     */
    protected function removeTables()
    {
    // redirector_redirects table
        $this->dropTableIfExists('{{%redirector_redirects}}');
    }
}


src/records/RedirectRecord.php
<?php
/**
 * Redirector plugin for Craft CMS 3.x
 *
 * A simple way to add 301/302 redirects within CraftCMS
 *
 * @link      https://github.com/jaetooledev
 * @copyright Copyright (c) 2021 Jae Toole
 */

namespace jaetooledev\redirector\records;

use jaetooledev\redirector\Redirector;

use Craft;
use craft\db\ActiveRecord;

/**
 * RedirectorRecord Record
 *
 * ActiveRecord is the base class for classes representing relational data in terms of objects.
 *
 * Active Record implements the [Active Record design pattern](http://en.wikipedia.org/wiki/Active_record).
 * The premise behind Active Record is that an individual [[ActiveRecord]] object is associated with a specific
 * row in a database table. The object's attributes are mapped to the columns of the corresponding table.
 * Referencing an Active Record attribute is equivalent to accessing the corresponding table column for that record.
 *
 * http://www.yiiframework.com/doc-2.0/guide-db-active-record.html
 *
 * @author    Jae Toole
 * @package   Redirector
 * @since     1.0.0
 */
class RedirectRecord extends ActiveRecord
{
    // Public Static Methods
    // =========================================================================

     /**
     * Declares the name of the database table associated with this AR class.
     * By default this method returns the class name as the table name by calling [[Inflector::camel2id()]]
     * with prefix [[Connection::tablePrefix]]. For example if [[Connection::tablePrefix]] is `tbl_`,
     * `Customer` becomes `tbl_customer`, and `OrderItem` becomes `tbl_order_item`. You may override this method
     * if the table is not named after this convention.
     *
     * By convention, tables created by plugins should be prefixed with the plugin
     * name and an underscore.
     *
     * @return string the table name
     */
    public static function tableName()
    {
        return '{{%redirector_redirects}}';
    }
}


src/services/RedirectService.php
<?php
/**
 * Redirector plugin for Craft CMS 3.x
 *
 * A simple way to add 301/302 redirects within CraftCMS
 *
 * @link      https://github.com/jaetooledev
 * @copyright Copyright (c) 2021 Jae Toole
 */

namespace jaetooledev\redirector\services;

use craft\helpers\Db;
use jaetooledev\redirector\records\RedirectRecord;
use jaetooledev\redirector\Redirector;

use Craft;
use craft\base\Component;

/**
 * RedirectService Service
 *
 * All of your plugin’s business logic should go in services, including saving data,
 * retrieving data, etc. They provide APIs that your controllers, template variables,
 * and other plugins can interact with.
 *
 * https://craftcms.com/docs/plugins/services
 *
 * @author    Jae Toole
 * @package   Redirector
 * @since     1.0.0
 */
class RedirectService extends Component
{
    // Public Methods
    // =========================================================================

    /**
     * This function can literally be anything you want, and you can have as many service
     * functions as you want
     *
     * From any other plugin file, call it like this:
     *
     *     Redirector::$plugin->redirectorService->exampleService()
     *
     * @return mixed
     */

    /**
     * Fetch all redirects for a given site.
     * @param int $siteId
     * @return array
     */
    public function getRedirectsForSite(int $siteId): array
    {
        return RedirectRecord::find()->where(['siteId' => $siteId])->all();
    }

    public function getRedirectById(int $redirectId): ?RedirectRecord
    {
        return RedirectRecord::findOne($redirectId);
    }

    /**
     * @throws \yii\db\StaleObjectException
     * @throws \Throwable
     */
    public function deleteRedirectById(int $redirectId): bool
    {
        return RedirectRecord::findOne($redirectId)->delete();
    }

    /**
     * Add a new redirect.
     * @param RedirectRecord $redirect
     * @return bool
     */
    public function save(RedirectRecord $redirect): bool
    {
       return $redirect->save();
    }
}


src/controllers/DashboardController.php
<?php

namespace jaetooledev\redirector\controllers;

use jaetooledev\redirector\records\RedirectRecord;
use jaetooledev\redirector\Redirector;

use Craft;
use craft\web\Controller;
use yii\web\BadRequestHttpException;
use yii\web\Response;

class DashboardController extends Controller
{

    public function actionIndex()
    {
        $redirects = Redirector::getInstance()->redirectService->getRedirectsForSite(
            Craft::$app->getSites()->currentSite->id
        );

        return $this->renderTemplate('redirector/index', [
            'redirects' => $redirects
        ]);
    }

    /**
     * @throws \yii\web\ForbiddenHttpException
     * @throws BadRequestHttpException
     */
    public function actionEdit(?int $redirectId = null, ?RedirectRecord $redirect = null): Response
    {
        $this->requirePermission('edit-redirects');
        if (!$redirect) {
            if ($redirectId) {
                $redirect = Redirector::getInstance()->redirectService->getRedirectById($redirectId);
                if (!$redirect) {
                    throw new BadRequestHttpException('Invalid Redirect ID: ' . $redirectId);
                }
            } else {
                $redirect = new RedirectRecord();
            }
        }

        $types = [
            '301' => '301',
            '302' => '302'
        ];

        return $this->renderTemplate('redirector/redirects/_edit', [
            'redirect' => $redirect,
            'types' => $types
        ]);
    }

    /**
     * @throws \yii\web\ForbiddenHttpException
     * @throws BadRequestHttpException
     */
    public function actionSave(): ?Response
    {
        $this->requirePermission('edit-redirects');
        $redirectId = $this->request->getBodyParam('redirectId');

        if ($redirectId) {
            $redirect = Redirector::getInstance()->redirectService->getRedirectById($redirectId);
            if (!$redirect) {
                throw new BadRequestHttpException('Invalid Redirect ID: ' . $redirectId);
            }
        } else {
            $redirect = new RedirectRecord();
        }

        $redirect->fromUrl = $this->request->getBodyParam('fromUrl');
        $redirect->toUrl = $this->request->getBodyParam('toUrl');
        $redirect->type = $this->request->getBodyParam('type');
        $redirect->siteId = Craft::$app->getSites()->currentSite->id;

        if (!Redirector::getInstance()->redirectService->save($redirect)) {
            if ($this->request->getAcceptsJson()) {
                return $this->asJson(
                    [
                        'errors' => $redirect->getErrors()
                    ]
                );
            }
            $this->setFailFlash(Craft::t('redirector', 'Error when saving Redirect'));

            Craft::$app->urlManager->setRouteParams(
                [
                    'redirect' => $redirect
                ]
            );
            return null;
        }

        if ($this->request->getAcceptsJson()) {
            return $this->asJson(
                [
                    'success' => true
                ]
            );
        }

        $this->setSuccessFlash(Craft::t('redirector', 'Redirect saved'));
        return $this->redirectToPostedUrl($redirect);
    }

    /**
     * @throws \Throwable
     * @throws \yii\db\StaleObjectException
     */
    public function actionDelete(int $redirectId)
    {
        if (!Redirector::getInstance()->redirectService->deleteRedirectById($redirectId)) { //TODO: Change this to allow better errors.
            if ($this->request->getAcceptsJson()) {
                return $this->asJson(
                    [
                        'errors' => [
                            Craft::t('redirector', 'Unable to delete Redirect')
                        ]
                    ]
                );
            }
            $this->setFailFlash(Craft::t('redirector', 'Error when deleting Redirect'));
            return null;
        }

        if ($this->request->getAcceptsJson()) {
            return $this->asJson(
                [
                    'success' => true
                ]
            );
        }

        $this->setSuccessFlash(Craft::t('redirector', 'Redirect deleted'));
        return $this->goBack('actions/redirector/dashboard/index');
    }
}
